File: product-service/internal/services/ws_product_service.go
package services

import (
	"context"
	"fmt"

	"github.com/PedroScheurer/product-service/internal/apperrors"
	"github.com/PedroScheurer/product-service/internal/dtos"
	"github.com/PedroScheurer/product-service/internal/entities"
	"github.com/PedroScheurer/product-service/internal/modules"
	"github.com/PedroScheurer/product-service/internal/repositories"
)

const adminType = 0

type WsProductService struct {
	repository repositories.ProductRepository
}

func NewWsProductService(repository repositories.ProductRepository) *WsProductService {
	return &WsProductService{repository: repository}
}

func (s *WsProductService) CreateProduct(ctx context.Context, dto dtos.ProductInDTO, userType int) (*dtos.ProductOutDTO, error) {
	if userType != adminType {
		return nil, apperrors.NewAuthenticationError("Usuário sem Permissão!")
	}

	product := &entities.ProductEntity{
		Name:         dto.Name,
		Instructor:   dto.Instructor,
		ImageURL:     dto.ImageURL,
		VideoURL:     dto.VideoURL,
		Description:  dto.Description,
		Workload:     dto.Workload,
		Modules:      modules.ResolveCount(dto.ModuleTitles, dto.Modules),
		ModuleTitles: modules.Encode(dto.ModuleTitles),
		Price:        dto.Price,
		Currency:     dto.Currency,
	}

	newProduct, err := s.repository.Save(ctx, product)
	if err != nil {
		return nil, fmt.Errorf("save product: %w", err)
	}

	return toProductOutDTO(newProduct), nil
}

func (s *WsProductService) AlterProduct(ctx context.Context, id int64, dto dtos.ProductUpdateInDTO, userType int) (*dtos.ProductOutDTO, error) {
	if userType != adminType {
		return nil, apperrors.NewAuthenticationError("Usuário sem Permissão!")
	}

	product, err := s.repository.FindByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("find product: %w", err)
	}
	if product == nil {
		return nil, apperrors.NewProductNotFoundError(fmt.Sprintf("Produto não encontrado com o ID: %d", id))
	}

	product.Name = dto.Name
	product.Instructor = dto.Instructor
	product.ImageURL = dto.ImageURL
	product.VideoURL = dto.VideoURL
	product.Description = dto.Description
	product.Workload = dto.Workload
	product.Modules = modules.ResolveCount(dto.ModuleTitles, dto.Modules)
	product.ModuleTitles = modules.Encode(dto.ModuleTitles)
	product.Price = dto.Price
	product.Currency = dto.Currency

	updatedProduct, err := s.repository.Save(ctx, product)
	if err != nil {
		return nil, fmt.Errorf("save product: %w", err)
	}

	return toProductOutDTO(updatedProduct), nil
}

func (s *WsProductService) DeleteProduct(ctx context.Context, id int64, userType int) error {
	if userType != adminType {
		return apperrors.NewAuthenticationError("Usuário sem Permissão!")
	}

	exists, err := s.repository.ExistsByID(ctx, id)
	if err != nil {
		return fmt.Errorf("check product exists: %w", err)
	}
	if !exists {
		return apperrors.NewProductNotFoundError(fmt.Sprintf("Produto não encontrado com o ID: %d", id))
	}

	if err := s.repository.DeleteByID(ctx, id); err != nil {
		return fmt.Errorf("delete product: %w", err)
	}

	return nil
}

func toProductOutDTO(product *entities.ProductEntity) *dtos.ProductOutDTO {
	return &dtos.ProductOutDTO{
		ID:           product.ID,
		Name:         product.Name,
		Instructor:   product.Instructor,
		ImageURL:     product.ImageURL,
		VideoURL:     product.VideoURL,
		Description:  product.Description,
		Workload:     product.Workload,
		Modules:      product.Modules,
		ModuleTitles: modules.Decode(product.ModuleTitles),
		Price:        product.Price,
		Currency:     product.Currency,
	}
}

File: product-service/src/main/java/br/edu/atitus/productservice/services/WsProductService.java
package br.edu.atitus.productservice.services;

import br.edu.atitus.productservice.dtos.ProductDTO;
import br.edu.atitus.productservice.dtos.ProductInDTO;
import br.edu.atitus.productservice.dtos.ProductOutDTO;
import br.edu.atitus.productservice.entities.ProductEntity;
import br.edu.atitus.productservice.infrastructure.exceptions.ProductNotFoundException;
import br.edu.atitus.productservice.repositories.ProductRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Service;

import javax.naming.AuthenticationException;

@Service
public class WsProductService {

    private final static Integer ADMIN_TYPE = 0;

    private final ProductRepository repository;

    public WsProductService(ProductRepository repository) {
        this.repository = repository;
    }

    public ProductOutDTO createProduct(ProductInDTO dto, Integer userType) throws AuthenticationException {
        if (userType.equals(ADMIN_TYPE))
            throw new AuthenticationException("Usuário sem Permissão!");

        ProductEntity product = convertDTO2Entity(dto);
        ProductEntity newProduct = repository.save(product);
        return new ProductOutDTO(newProduct.getId(),
                newProduct.getDescription(),
                newProduct.getBrand(),
                newProduct.getModel(),
                newProduct.getPrice(),
                newProduct.getCurrency(),
                newProduct.getStock());
    }


    public ProductOutDTO alterProduct(Long idProduct, ProductDTO dto, Integer userType) throws AuthenticationException {
        if (!userType.equals(ADMIN_TYPE)) {
            throw new AuthenticationException("Usuário sem Permissão!");
        }

        ProductEntity product = repository.findById(idProduct)
                .orElseThrow(() -> new ProductNotFoundException("Produto não encontrado com o ID: " + idProduct));

        product.setDescription(dto.description());
        product.setBrand(dto.brand());
        product.setModel(dto.model());
        product.setPrice(dto.price());
        product.setCurrency(dto.currency());

        ProductEntity updatedEntity = repository.save(product);

        return new ProductOutDTO(
                updatedEntity.getId(),
                updatedEntity.getDescription(),
                updatedEntity.getBrand(),
                updatedEntity.getModel(),
                updatedEntity.getPrice(),
                updatedEntity.getCurrency(),
                updatedEntity.getStock()
        );
    }

    public void deleteProduct(Long idProduct, Integer userType) throws AuthenticationException {
        if (!userType.equals(ADMIN_TYPE)) {
            throw new AuthenticationException("Usuário sem Permissão!");
        }

        if (!repository.existsById(idProduct)) {
            throw new ProductNotFoundException("Produto não encontrado com o ID: " + idProduct);
        }
        repository.deleteById(idProduct);
    }

    private ProductEntity convertDTO2Entity(ProductInDTO dto){
        var entity = new ProductEntity();
        BeanUtils.copyProperties(dto, entity);
        return entity;
    }


}
